Add `url_rules` option to HTML diff

ftsalamp opened the pull request:
This PR aims to fix the false positive issue that appears when comparing webpages whose resources might have extra information. This information might be a timestamp in the path of a linked resource (occurring when diffing web archives of pages) or a jsessionid, etc.

In this PR I am implementing the __eq__ functions on the href_token and tag_token classes.
If the URL parameter "strict_urls" is set, and there is a matching value in the REGEX_RULES dictionary of the Strict_url_rule class, the __eq__ function of those objects will take into account the appropriate rule when deciding if the objects are equal or not.

This solves issue #391 

Looking forward to any comments/suggestions on my approach :)

vbanos commented on web_monitoring/html_diff_render.py:
```diff
@@ -187,8 +187,16 @@
+    mode = False
+    REGEX_RULES = {'WBM': r"/web/\d{14}",
+                   'jsessionid': r";jsessionid[^;=]*=([^;]+)",
+                   'UKWA': r"https://www\.webarchive\.org\.uk/wayback/en/archive/\d{14}mp_/"}
```
Performance improvement idea:

Use `re.compile(r"/web/\d{14}")` here to compile regex initially and store them in the `REGEX_RULES` dict.

Then, later do:
```
        match_element = Strict_url_rule.REGEX_RULES[Strict_url_rule.mode].search(element)
```
Instead of what you currently do:
```
        match_element = re.search(Strict_url_rule.REGEX_RULES[Strict_url_rule.mode], element)
```
It will be MUCH faster.

Mr0grog commented on web_monitoring/html_diff_render.py:
```diff
@@ -681,6 +714,21 @@ def fixup_chunks(chunks):
+        return str.__eq__(element, other) or str.__eq__(element.lower(), other.lower())
+    except KeyError:
+        raise KeyError("{} is an invalid strict URL rule."
+                       "".format(Strict_url_rule.mode))
```
You could simplify this with an f-string:

```py
raise KeyError(f'{StrictUrlRule.mode} is an invalid strict URL rule.')
```

ftsalamp replied:
This is how a KeyError is raised in diffing_server.py but your suggestion is cleaner.

Mr0grog commented on web_monitoring/html_diff_render.py:
```diff
@@ -536,6 +549,16 @@ def __new__(cls, tag, data, html_repr, pre_tags=None,
+        # This equality check aims to apply specific rules to the contents
+        # of the tag element solving false positive cases
+        if Strict_url_rule.mode:
+            return clean_resource_url(self, other)
```
I think this might wind up being way too broad. First, if `other` is a normal word in the page’s text (perhaps a URL that is fully written out instead of an actual element), you’re going to wind up potentially considering this tag equal to it, which clearly isn’t right.

You should probably start by making sure `other` is also an instance of `tag_token`.

Then it might be a good idea to try and pull out all the URLs or URL-containing attributes to compare. For example, an `<img ...>` tag might have `src`, but it might also have `srcset`.

Then, since this is a generic comparison instead of one specific to `<img ...>` tags, you should probably also compare all the other attributes if the URL-containing ones are equal. (Alternatively, you could create a new `ImgToken` type that inherits from `tag_token` and is just for `<img ...>` tags and implement `__eq__` on that instead.)

Mr0grog commented on web_monitoring/html_diff_render.py:
```diff
@@ -681,6 +714,21 @@ def fixup_chunks(chunks):
+        match_other = Strict_url_rule.REGEX_RULES[Strict_url_rule.mode].search(other)
+        if match_other:
+            other = other[match_other.end():]
+        return str.__eq__(element, other) or str.__eq__(element.lower(), other.lower())
```
What if these elements contain multiple things that look like URLs in them, or contain the same URLs in different orders? Consider that an `<img>` elements `alt` attribute could contain a URL, or that it might have multiple URLs in its `srcset`, or that a refactor of the page might have caused `src` and `srcset` attributes to be in a different order!

All these considerations might be different for different elements, so one improvement here might be to make this function take the *URLs* to compare, rather than the elements, and leave extracting the URLs to the token class, which can be customized for different elements or types of content.

Mr0grog commented on web_monitoring/html_diff_render.py:
```diff
@@ -187,8 +187,16 @@
+    mode = False
+    REGEX_RULES = {'WBM': re.compile(r"/web/\d{14}"),
+                   'jsessionid': re.compile(r";jsessionid[^;=]*=([^;]+)"),
+                   'UKWA': re.compile(r"https://www\.webarchive\.org\.uk/wayback/en/archive/\d{14}mp_/")}
```
Could you please include comments or docstrings describing what each of these are meant to do? We both know the purpose here, but it might not be clear to someone else that:

- `WBM` stands for Wayback Machine
- The use-case here is in comparing Wayback mementos, where links will have been rewritten to point to another memento in Wayback instead of the original URL.
- etc.

^ Same sort of issues for each rule here, not just `WBM`. I know why we need to consider `jsessionid`, but someone else might not.

ftsalamp replied:
Added

Mr0grog commented on web_monitoring/html_diff_render.py:
```diff
@@ -681,6 +714,21 @@ def fixup_chunks(chunks):
+    try:
+        match_element = Strict_url_rule.REGEX_RULES[Strict_url_rule.mode].search(element)
+        if match_element:
+            element = element[match_element.end():]
```
I am not sure this is quite right. Looking at everything after the match makes sense for Wayback URLs, but for the `jsessionid` rule, where we only want to cut that one bit of text out of the middle of the URL. There is important URL text before it and potentially an important querystring after it.

A better approach would probably be to have different comparison/rule classes (e.g. one for Wayback, one for UK Wayback, one for Java servlet sessions) that each implement their own logic for comparing two URLs. When you set up the diff, you choose which classes to use when comparing. For example:

```py
class WaybackUrlComparator:
    matcher = re.compile(r'/web/\d{14}')

    def compare(self, url_a, url_b):
        match_a = self.matcher.search(url_a)
        match_b = self.matcher.search(url_b)
        if match_a and match_b:
            return url_a[match_a.end():] == url_b[match_b.end():]
        else:
            return url_a == url_b

class WaybackUkUrlComparator(WaybackUrlComparator):
    matcher = re.compile(r'https://www\.webarchive\.org\.uk/wayback/en/archive/\d{14}mp_/')

class ServletSessionUrlComparator:
    matcher = re.compile(r';jsessionid=[^;]+')

    def compare(self, url_a, url_b):
        match_a = self.matcher.sub('', url_a)
        match_b = self.matcher.sub('', url_b)
        return match_a == match_b

ftsalamp replied:
Done

ftsalamp commented on web_monitoring/html_diff_render.py:
```diff
@@ -536,6 +549,16 @@ def __new__(cls, tag, data, html_repr, pre_tags=None,
+        return str.__eq__(self, other) or str.__eq__(self.lower(), other.lower())
+
+    def __hash__(self):
+        return hash(self.lower())
```
To be honest I got carried away from the __hash__ implementation on the Link class in links_diff.py

I've corrected it!

ftsalamp commented on web_monitoring/html_diff_render.py:
```diff
@@ -536,6 +549,16 @@ def __new__(cls, tag, data, html_repr, pre_tags=None,
+        # of the tag element solving false positive cases
+        if Strict_url_rule.mode:
+            return clean_resource_url(self, other)
+        return str.__eq__(self, other) or str.__eq__(self.lower(), other.lower())
```
Same reason as below. Sorry about this!

ftsalamp commented on web_monitoring/html_diff_render.py:
```diff
@@ -255,6 +266,8 @@ def html_diff_render(a_text, b_text, a_headers=None, b_headers=None,
         b_headers,
         content_type_options)
 
+    Strict_url_rule.mode = strict_urls
```
I added the parameter as this solution seems ok for now.

ftsalamp commented on web_monitoring/html_diff_render.py:
```diff
@@ -187,8 +187,16 @@
 def html_diff_render(a_text, b_text, a_headers=None, b_headers=None,
-                     include='combined', content_type_options='normal'):
+                     include='combined', content_type_options='normal',
+                     strict_urls=False):
```
You're right!

ftsalamp commented on web_monitoring/html_diff_render.py:
```diff
@@ -241,6 +249,9 @@ def html_diff_render(a_text, b_text, a_headers=None, b_headers=None,
         - `ignore` doesn’t do any checking at all.
+    strict_urls : string
+        The value of the URL parameter that indicates if the tag elements and href
+        elements should use special rules when checking equality,
```
Done! :)

ftsalamp commented on web_monitoring/html_diff_render.py:
```diff
@@ -187,8 +187,16 @@
 MAX_SPACERS = 2500
 
 
+class Strict_url_rule:
```
Fixed!

ftsalamp commented on web_monitoring/html_diff_render.py:
```diff
@@ -681,6 +714,21 @@ def fixup_chunks(chunks):
     return result
 
+
+def clean_resource_url(element, other):
```
Moved

Mr0grog commented on web_monitoring/html_diff_render.py:
```diff
@@ -544,15 +635,38 @@ def __repr__(self):
+        return super().__eq__(other)
+
+    def __hash__(self):
+        return super.__hash__(self)
```
I think you can just leave this method out; the effect will be the same.

Mr0grog commented on web_monitoring/html_diff_render.py:
```diff
@@ -810,17 +926,51 @@ def html(self):
 class ImgTagToken(tag_token):
-    def __new__(cls, tag, data, html_repr, pre_tags=None,
+
+    regex = re.compile(r"(src|srcset)=('|\")(.*?)('|\")")
```
I don’t think you actually need this at all (see below), but if you keep it, please use a more descriptive name.

Mr0grog commented on web_monitoring/html_diff_render.py:
```diff
@@ -810,17 +926,51 @@ def html(self):
+
+    def are_any_urls_equal(self, other):
+        strict_urls = self.strict_urls
+        self_array = self[7:-3]
```
Once we get to this point, we should hopefully not be doing any more parsing of strings. I think you should probably be modifying `flatten_el`, `fixup_chunks`, and `_customize_token` here. See below for more.

Mr0grog commented on web_monitoring/html_diff_render.py:
```diff
@@ -810,17 +926,51 @@ def html(self):
+    def __new__(cls, tag, data, html_repr, strict_urls, pre_tags=None,
                 post_tags=None, trailing_whitespace=""):
-        obj = DiffToken.__new__(cls, "\n\nImg:%s\n\n" % data,
+        found_urls = ImgTagToken.regex.findall(html_repr)
```
In `flatten_el`, we grab the actual `src` attribute value out of the `img` element: https://github.com/edgi-govdata-archiving/web-monitoring-processing/blob/e602b3f94cdcbebc54db0861b38aa6f2a665b9bc/web_monitoring/html_diff_render.py#L693-L694

And then in `fixup_chunks`, we create a `tag_token` instance` where the `data` is just the `src` attribute: https://github.com/edgi-govdata-archiving/web-monitoring-processing/blob/e602b3f94cdcbebc54db0861b38aa6f2a665b9bc/web_monitoring/html_diff_render.py#L636-L641

(So at this point, we’ve already parsed out the src and have it available.)

Finally, in `_customize_token`, we replace the `tag_token` instance with a `ImgTagToken`: https://github.com/edgi-govdata-archiving/web-monitoring-processing/blob/e602b3f94cdcbebc54db0861b38aa6f2a665b9bc/web_monitoring/html_diff_render.py#L1036-L1044

A few things are important to note here:

- We already have the value of the `src` element in the form of `self.data`, so there’s no need to parse it out again here. We should just compare `self.data` to `other.data` when we compare.

- When parsing in `flatten_el`, we already have meaningful data in the form of `etree`, so we generally try and pull out all the important stuff there instead of doing more string parsing. At current, we get the `src` that way, but we could also get `srcset` if we want to use it in comparisons. That way we don’t have to parse it again here.

- We only do this messy middle step with `tag_token` because of the incremental way this was forked from lxml. Ideally, we’d refactor to get rid of `_customize_token` entirely. If you want to extract `srcset` for use now, you should…

    1. Modify `flatten_el` to pull that info out.
    2. Modify `fixup_chunks` to create a `ImgTagToken` instance with that data instead of making a `tag_token`.
    3. Then you should get rid of the logic in `_customize_token` that makes `ImgTagToken` instances, because it would never run anymore.
    4. Clean up the logic here and in the `are_any_urls_equal` method to use that data instead of doing more string parsing.

Mr0grog commented on web_monitoring/html_diff_render.py:
```diff
@@ -810,17 +926,51 @@ def html(self):
+        return False
+
+    def __hash__(self):
+        return super.__hash__(self)
```
I think you can just leave this method out; the effect will be the same.

ftsalamp replied:
If I don't include this I get a `TypeError: unhashable type: 'ImgTagToken'`

Mr0grog replied:
Ah! I hadn’t realized Python has some special behavior when `__eq__` is defined. From [the docs](https://docs.python.org/3/reference/datamodel.html#object.__hash__):

> A class that overrides `__eq__()` and does not define `__hash__()` will have its `__hash__()` implicitly set to `None`… If a class that overrides `__eq__()` needs to retain the implementation of `__hash__()` from a parent class, the interpreter must be told this explicitly by setting `__hash__ = <ParentClass>.__hash__`.

So I guess you do need to implement `__hash__`, but it might be better to use the suggested for above, or to at least change `super.__hash__(self)` to use the more standard `super().__hash__()` form.

Mr0grog commented:
@ftsalamp This could also use some tests, please!

Mr0grog commented on web_monitoring/html_diff_render.py:
```diff
@@ -187,8 +187,74 @@
+    def clean_resource_url(element, other, mode):
+        try:
+            comparator = StrictUrlRule.CLASS_RULES[mode]()
+            return comparator.compare(element, other)
```
> Additionally, you probably want a way to combine the comparators here

To be clear, the *combining* stuff doesn’t need to be in this PR! We can definitely do that later as a separate piece of work.

ftsalamp commented on web_monitoring/html_diff_render.py:
```diff
@@ -187,8 +187,74 @@
 
+class WaybackUrlComparator:
+    matcher = re.compile(r'web/\d{14}(im_)?/(https?://)?(www.)?')
+    url_schema_matcher = re.compile(r'(https?\:\/\/)?(www\.)?')
```
You are correct! The `url_schema_matcher` is redundant.

ftsalamp commented on web_monitoring/html_diff_render.py:
```diff
@@ -187,8 +187,74 @@
 
 
+class WaybackUrlComparator:
+    matcher = re.compile(r'web/\d{14}(im_)?/(https?://)?(www.)?')
```
> Finally, in the ending part, is there a case where it would not be an absolute URL? i.e. is (https?://)? really optional? If not, we should probably also remove the (www.)? after it, since that isn’t really doing anything either. Should this also be restricted to http, either? Couldn’t some of these URLs be ftp://, etc?

We could have http://www.anexample......
                        http://anexample......
                        https://www.anexample......
                        https://anexample......
So I believe the regex is correct.
I haven't run into any `ftp://` 's etc so far. @vbanos is this a possibility?

ftsalamp replied:
> Based on that, it looks it should probably be (id_|im_|js_|cs_)?.
There can't be _id in a snapshot but the rest should be added.

Mr0grog replied:
So, based on that, it looks like the protocol, at least, is not optional, so you should not have `(https?://)?`, you should have `https?://`.

ftsalamp replied:
Sorry, I omitted the examples with the protocol. Again it might or might not be there.

ftsalamp commented on web_monitoring/html_diff_render.py:
```diff
@@ -187,8 +187,74 @@
+
+
+class WaybackUkUrlComparator(WaybackUrlComparator):
+    matcher = re.compile(r'https://www\.webarchive\.org\.uk/wayback/en/archive/\d{14}mp_/')
```
> Is mp_ a valid suffix that’s not documented (or documented somewhere I don’t know about)? Is it particular to the UK?

I didn't find any documentation about it but when looking at the HTML code UKWA returns, if a link has the `_mp` attribute, the "raw" snapshot as it was saved is returned. And all their links have it. Otherwise, the source link is rendered in an iframe using JS, and is thus undiffable.

It it specific to the UKWA.

Mr0grog commented:
@ftsalamp do you think you might be able to rebase this whole branch (it seems like you rebased in the middle somehow) so we get a more readable diff and don’t rewrite a bunch of commits that are already on master?

ftsalamp commented:
@Mr0grog I've added some tests and fixed a bug they uncovered. If you think we are ready to merge I'll make a new PR from a new branch created from master with a cleaner (only two commits) commit history. Let's discuss the changes here and close the PR when we don't need it anymore.

Mr0grog commented:
Rebased this branch so I can see what’s actually happening here in order to review.

Mr0grog commented on web_monitoring/tests/test_html_diff_validity.py:
```diff
@@ -175,3 +175,85 @@ def test_html_diff_works_on_documents_with_no_body():
+        '<a href="https://www.ncdc.noaa.gov/homr/api;jsessionid=A2DECB66D2648B'
+        'FED11FC721FC3043A1"></a>',
+        '<a href="https://www.ncdc.noaa.gov/homr/api;jsessionid=45312D9542FDB0'
+        '15289A1BBD76958F43"></a>',
```
Same formatting note as above. It’s OK to go over line length if it’s easier to read.

Mr0grog commented on web_monitoring/tests/test_html_diff_validity.py:
```diff
@@ -175,3 +175,85 @@ def test_html_diff_works_on_documents_with_no_body():
+        '20190525141538/https://www.example.gov/></a>',
+        '<a href="https://www.webarchive.org.uk/wayback/en/archive/'
+        '20181231224558/https://www.example.gov/></a>',
+        include='all', strict_urls='UK_WBM')
```
Same formatting note as above.

Mr0grog commented on web_monitoring/tests/test_html_diff_validity.py:
```diff
@@ -175,3 +175,85 @@ def test_html_diff_works_on_documents_with_no_body():
     assert isinstance(results['combined'], str)
+
+
+def test_html_diff_works_wbm_snapshots():
```
Would you please also add at least one test case for the *non-equal* situation? For example, the exact same test as this, but with `strict_urls=None` and asserting that `results['change_count'] == 1`.

Mr0grog commented:
Rebased on master again. Working on updating and finalizing this today.

Mr0grog commented:
OK, going ahead and merging this. Will make separate follow-up issues for comparison vs. normalization and for applying this logic to the links diff.

vbanos commented on web_monitoring/tests/test_html_diff_validity.py:
```diff
@@ -175,3 +175,127 @@ def test_html_diff_works_on_documents_with_no_body():
+        ''',
+        include='all')
+
+    assert results['change_count'] == 0
```
`srcset` can contain various images. The decision on which one to load depends on the current client resolution.
I agree that ALL need to match, not just one.
